src/media-server.js

```diff
@@ -0,0 +1,148 @@
+import { once } from 'events'
+import fastify from 'fastify'
+import pTimeout from 'p-timeout'
+import StateMachine from 'start-stop-state-machine'
+
+import BlobServerPlugin from './fastify-plugins/blobs.js'
+import { kBlobStore } from './mapeo-project.js'
+
+export const BLOBS_PREFIX = 'blobs'
+export const ICONS_PREFIX = 'icons'
+
+/**
+ * @typedef {Object} StartOpts
+ *
+ * @property {string} [host]
+ * @property {number} [port]
+ */
+
+export class MediaServer {
+  #serverState
+  #fastify
+  #createFastify
+
+  /**
+   * @param {object} params
+   * @param {(projectPublicId: string) => Promise<import('./mapeo-project.js').MapeoProject>} params.getProject
+   * @param {import('fastify').FastifyServerOptions['logger']} [params.logger]
+   */
+  constructor({ getProject, logger }) {
+    this.#createFastify = () => {
+      const server = fastify({ logger })
+
+      server.register(BlobServerPlugin, {
+        prefix: BLOBS_PREFIX,
+        getBlobStore: async (projectPublicId) => {
+          const project = await getProject(projectPublicId)
+          return project[kBlobStore]
+        },
+      })
+
+      return server
+    }
+
+    this.#fastify = this.#createFastify()
+
+    this.#serverState = new StateMachine({
+      start: this.#startServer.bind(this),
+      stop: this.#stopServer.bind(this),
+    })
+  }
+
+  /**
+   * @param {StartOpts} [opts]
+   */
+  async #startServer({ host = '127.0.0.1', port } = {}) {
+    await this.#fastify.listen({ host, port })
+  }
+
+  async #stopServer() {
+    await this.#fastify.close()
+  }
+
+  /**
+   * @returns {Promise<string>}
+   */
+  async #getAddress() {
+    return pTimeout(getServerAddress(this.#fastify.server), {
+      milliseconds: 1000,
```

Comment:
not sure what the timeout should be here

Reply:
I think that's reasonable